elf_reader.go

```diff
@@ -279,86 +282,139 @@ func (ec *elfCode) loadPrograms() (map[string]*ProgramSpec, error) {
+			if ec.btf != nil {
+				spec.BTF, err = ec.btf.Program(name)
+				if err != nil && !errors.Is(err, btf.ErrNoExtendedInfo) {
+					return nil, fmt.Errorf("program %s: %w", name, err)
+				}
+			}
+
+			// Function names must be unique within a single ELF blob.
+			if progs[name] != nil {
+				return nil, fmt.Errorf("duplicate program name %s", name)
 			}
+			progs[name] = spec
 		}
+	}
 
-		if spec.Type == UnspecifiedProgram {
-			// There is no single name we can use for "library" sections,
-			// since they may contain multiple functions. We'll decode the
-			// labels they contain later on, and then link sections that way.
-			libs = append(libs, spec)
-		} else {
-			progs = append(progs, spec)
+	// Populate each program's neighbour table with pointers to the other
+	// programs it directly references.
+	if err := findNeighbours(progs); err != nil {
+		return nil, fmt.Errorf("finding program neighbours: %w", err)
+	}
+
+	// Don't emit programs of unknown type to preserve backwards compatibility.
+	for n, p := range progs {
+		if p.Type == UnspecifiedProgram {
+			delete(progs, n)
 		}
 	}
 
-	res := make(map[string]*ProgramSpec, len(progs))
-	for _, prog := range progs {
-		err := link(prog, libs)
+	return progs, nil
+}
+
+// loadFunctions extracts instruction streams from the given program section
+// starting at each symbol in the section. The section's symbols must already
+// be narrowed down to STT_NOTYPE (emitted by clang <8) or STT_FUNC.
+//
+// The resulting map is indexed by function name.
+func (ec *elfCode) loadFunctions(section *elfSection) (map[string]asm.Instructions, error) {
+	funcs := make(map[string]asm.Instructions)
+	for _, sym := range section.symbols {
+		// Load the function's instructions. Pass in the absolute offset of the
+		// function within the section to allow resolving relocations.
+		si, err := ec.loadInstructions(section, sym)
 		if err != nil {
-			return nil, fmt.Errorf("program %s: %w", prog.Name, err)
+			return nil, fmt.Errorf("loading instructions: %w", err)
 		}
-		res[prog.Name] = prog
+
+		funcs[sym.Name] = si
 	}
 
-	return res, nil
+	return funcs, nil
 }
 
-func (ec *elfCode) loadInstructions(section *elfSection) (asm.Instructions, uint64, error) {
-	var (
-		r      = bufio.NewReader(section.Open())
-		insns  asm.Instructions
-		offset uint64
-	)
+// loadInstructions extracts the instruction stream of the given symbol
+// from the given ELF section.
+func (ec *elfCode) loadInstructions(section *elfSection, sym elf.Symbol) (asm.Instructions, error) {
+	// clang up until at least version 7 does not set the symbol's size field.
+	// In this case, buffer the remainder of the section, since we only read up
+	// to the first Exit instruction.
+	if sym.Size == 0 {
+		// Don't use Max(U)Int64, the offset is added to this number
+		// during SectionReader creation and will overflow.
+		sym.Size = section.Size
+	}
+
+	r := internal.NewBufferedSectionReader(section, int64(sym.Value), int64(sym.Size))
```

Comment:
> This ends up recreating the buffer for every function in a section,

`section` is `io.ReaderAt` backed by a file or memory, we never double-buffer or read a function more than once generally. It allocates one buffer per function, which sounds reasonable at this point.

>  I thought you want to allocate less?

I assume you're comparing this to our discussion regarding the type graph from a few weeks ago. A few very key differences there:

- this concerns several orders of magnitude fewer allocs than the vmlinux type graph
- these buffers are generally large allocs, and
- they are always accessed, as opposed to the majority of the vmlinux type graph that is allocated and simply never accessed afterwards

The primary goal of this BufferedSectionReader is to cleanly express a read at a certain offset and length while keeping the amount of syscalls down as to not impact 9pfs performance.

> Why not wrap the section in a bufio.Reader and pass that to loadInstructions?

`section.symbols` is a map, so functions are parsed in random order, and `bufio.Reader` does not provide the random access needed for this. Do you mean we should create a single buffer at the start of `loadFunctions()` that `loadInstructions()` simply creates a `SectionReader` into, like:

```
func loadFunctions() {
...
sd, _ := section.Data()
ra := bytes.NewReader(sd)
...
    ec.loadInstructions(ra, sym)
```
?

We need `*elfSection` for multiple other purposes in `loadInstructions()`, so IMO not a worthwhile optimization and won't improve readability. I would probably keep this part as-is.

> Follow up question: do functions have to be consecutive?

What do you mean?

Reply:
The gist of my question is: instead of iterating symbols and parsing functions in random order, iterate instructions and look up function symbols. That would allow us doing a linear read using a single buffer and makes it easy to verify that there are no "gaps" between functions.

Reply:
Sure, that's worth exploring. Would also allow us to get rid of the <clang7 workaround since it doesn't set the symbol size correctly.

Reply:
Done!